### DISCUSSION TITLE
Lab rework

### DISCUSSION THREAD

**Discussion by curstwist** [CleverRaven/Cataclysm-DDA#46500] [Ideas]
I've begun some planning & drafting for a full lab overhaul after 0.F stable is released.  These are by far the most played maps in the game so I wanted to solicit feedback and ideas before I get too far.

The biggest change that needs to happen is procgen labs are replaced by full json created and spawned labs which will allow for consistent z level transitions.  Labs are currently the biggest barrier to removing the old procgen stairs.  

My over all goals include
- more architecturally sensible labs (residential, lab, security, etc will all be distinct areas).
- labs will be thematic (mutagen labs, weapons labs, bionics labs, portal labs etc).
- more traps and situational problem solving instead of the randomized trap rooms and hazards.
- areas with increasing difficulty, access restrictions
- randomization through nested chunks and variants instead of the current randomized mapgen.

There are two approaches I can see taking for the next generation of labs.  The first is by using something similar to the microlab structures that mimic old labs fairly closely.  This approach will make segmenting lab areas more difficult and will probably require them to be themed by z level only.  The z level transitions will be more difficult to make feel different (they will always be in the same spots).

The second approach is more complex but is basically the creation of a variety of "mini" specials and combining them into map specials.  So you can have different multi-level residential areas that combine with different security or laboratory themed specials.  This approach will take the longest to create enough content to get a random feel to the maps again.  They may feel quite similar outside of the lab research areas until enough variants are made.  

No matter the approach,  I do plan on re-imagining a lot of the trap usage and rely more on traps, security checkpoints and other more realistic approaches to security for a top secret facility.

Some of the perks of moving labs to full json will include:
- better z level transitions including open air/vaulted ceiling areas if I use the module approach.
- integration of vehicle ramps, elevators and stairs.
- new traps including some update_mapgen traps.
- overall better looking labs.
- new monsters and in particular some new bots to support labs.
- re-evaluating loot.

Overall, I'm leaning towards the 2nd approach of lab modules that fit into larger specials but I'd like some input, in particular from other lab & map content creators. 

Once I pin down the approach, I'll probably start a project page to track it since I'll need to roll it out in phases.  

Feel free to leave thoughts on what you'd like to experience in labs, thematic areas they should contain, and what elements should be preserved from the old labs.

**Comment by anothersimulacrum** [comment]
> The second approach is more complex but is basically the creation of a variety of "mini" specials and combining them into map specials. So you can have different multi-level residential areas that combine with different security or laboratory themed specials. This approach will take the longest to create enough content to get a random feel to the maps again. They may feel quite similar outside of the lab research areas until enough variants are made.

I'm a bigger fan of this approach. It gives more freedom and allows for a better overall implementation. I am not sure we really need to stick close to what labs started as.

**Comment by Maleclypse** [comment]
I definitely support the second option.  Do we know or have an idea of how mods would be able to add mini-specials in to this?

Edit:  So I'm thinking if I created like a 2x, 2y, 2z barracks would it have to find a matching vanilla 2x, 2y, 2z minispecial? Or could I just `"extend"` in to the OM Special randomizer.

**Reply by curstwist** [reply]
- One way would be like the basements, where you can use nested map chunks with the same name.    
- 2nd way would be using the same OMT names as an existing mini-special.  I'm still experimenting but I'd like the names to be based on size and function.  Then you'd just want to match any entrance and exit points from the vanilla special.
- 3rd way would be to make your own modular special pulling what you want from vanilla and inserting your own ones as needed.

There may be other ways but those all seem pretty versatile depending on the amount of control you want.  I'm not entirely sure how to handle overmap spawning rates yet because we'd be effectively increasing the number of labs in the spawn lists.

Edit: I haven't looked into extends much but that would be a nice approach.  I want to test how else_chunks work too for extending mapgen and if it'll work with pure json labs.

**Comment by actual-nh** [comment]
Ideally, some of the lab submodules would be usable other places - hospitals and college/university labs are the most obvious, but something like a security section might be usable in a jail or other secure facility. (A local item blacklist may be needed, admittedly.)

**Reply by curstwist** [reply]
yeah, these are all designed to be underground but reuse is possible, the nested map selection may make it tricky as well.  We need a better way to browse existing nested map chunks (debug spawn helps a lot but it's a little tedious).

**Reply by actual-nh** [reply]
Understand. (In terms of underground - a research hospital I was around for a while had the morgue in the basement, and I would also not be surprised to see high-dose radiation therapy - for cancer - located there. Also, any lab with an electron microscope may well be underground to reduce vibrations.)

**Comment by Venera3** [comment]
Very much in favor of moving labs into a less gamey direction. I see quite a few problems with them as they stand currently:

They offer no gameplay challenges you don't encounter outside, so you can essentially play the same "peek-fight-run" game you play in a city, only with extra darkness and many doors making them less challenging than they should be. A heavier reliance on traps/forcing you to engage X mob for a keycard or similar would help with that tremendously.

They are too common. At the moment there are around a dozen super-duper secret XEDRA blacksites on all overmaps, making them the seemingly most important employer in New England. That means you don't need to bother with a challenging lab, since you can just hop on over to the next one and try your luck there, and leading to an abundance of supposedly high-tech loot that propels any character that's lab-ready ( and an M4A1 and a few grenades suffice for most ) into lategame gear without spending any time in the midgame. Labs being bigger, rarer and more focused on one particular field would help slow progression and pad out the interesting part of the game.

Because of the way map specials are generated on worldgen, any of the Z-levels having a lab disqualifies that overmap chunk from spawning another special, making the surface feel barren (and incidentally leaving you with a pretty good guess where labs are). I'm going to try and adjust that part as well I can, but I'm not sure how far I can push it without breaking the whole "subway-only central lab" concept.

**Reply by actual-nh** [reply]
About to go back to work, but - is the problem a special potentially having Z-levels that would run into the lab subways? If some mechanism could filter out specials that could have such, that could fix the problem. (Even if there wasn't an active mechanism checking for potential Z-levels from a special, tagging those that don't so they could still be used would help, and avoid incompatibility with mods, etc.)

**Reply by GyprockGypsy** [reply]
I vote type 2. Can't really offer much in the way of technical help, but IMO it would nice to see the labs be based off of a central corridor design, with security rooms/features placed along the way and checkpoints at either end. Or a column design with increasing levels of security as you go down. Anything is better than finding the next stairs down inside a high security specimen containment with a Mi-Go. How would anyone have ever worked in this place? OSHA would have a fucking field day.

**Comment by tenmillimaster** [comment]
> areas with increasing difficulty, access restrictions

Very much this, please. The labs should be compartmentalized, with entire teams of personnel having no idea what the actual purpose of the building is.

**Reply by actual-nh** [reply]
> > areas with increasing difficulty, access restrictions
> 
> Very much this, please. The labs should be compartmentalized, with entire teams of personnel having no idea what the actual purpose of the building is.

Or, at least, initially not having known. Some interesting snippets should be extractable from their discovering things...

**Comment by curstwist** [comment]
So I've got a good start on the "mini-specials" approach that everyone here supports.  It's going to require an enormous amount of content of which maps is a large part.

Other areas we'll need content are monsters, special lab boss monsters, NPC's to encounter (long term plan), new items, special loot items, vehicles, terrains, furniture, traps, snippet content.

If anyone wants to be involved in content creation or willing to work on features to support new game play for labs can reach out to me.  I will see about getting a project page started with my intentions and design thoughts to date.

Right now, all my WIP content is confined to a mod so I can easily share it and work on it.

**Reply by Xaleth** [reply]
I can help with that. Just let me know where to start

**Reply by GyprockGypsy** [reply]
Could anyone point me toward a wiki or how-to for making the maps? I've been mapping since AD&D, so the 1 meter tile is pretty familiar to me (well, its a 2.5 yard tile in AD&D, but whatever). Maybe we could also turn this into a resource for people who want learn how to map in Json if there isn't one.

**Reply by actual-nh** [reply]
> Could anyone point me toward a wiki or how-to for making the maps?

I would start with [here](https://github.com/CleverRaven/Cataclysm-DDA/tree/master/doc/JSON%20Mapping%20Guides), then take a look at [doc/MAPGEN.md](https://github.com/CleverRaven/Cataclysm-DDA/tree/master/doc/MAPGEN.md) and possibly [doc/OVERMAP.md](https://github.com/CleverRaven/Cataclysm-DDA/tree/master/doc/OVERMAP.md).
